HP-908 remove useless profile exists check

NikoHelle opened the pull request:
The existence of a profile can be checked just by loading the user's profile. So the profile exists check is skipped now and removed completely.

Checking if user...
- is authenticated
- is loading a profile 
- has a profile 
- profile is loaded, but does not exist
- profile load failed
- authentication failed

...is a bit complicated. And due the nature of a SPA, user may navigate back and forth after any loading has started. Which makes checks even more complicated.

So added a hook which tracks the different scenarios reliably and with comprehensive tests.

NikoHelle committed:
(HP-908) Add new mocks to MockApolloClient

Added ability to return allowed graphQL-error in response along with profile data.

Also made a helper to create that error with specific props.

Added ability to return response data for createProfile

NikoHelle committed:
(HP-908) Fix bug in getErrorMessage

The returned networkError may be an object. In that case return its body-property which is a string

NikoHelle committed:
(HP-908) Make a generic exposeHook helper function

Converted exposeProfileMutationsHook to be a generic hook helper.

Added "waitForProfileData" argument, because other hooks may not need profile to exist.

NikoHelle committed:
(HP-908) Extracted create profile form submit flow to re-usable function

The flow is needed in Profile tests

NikoHelle committed:
(HP-908) Removed ProfileExistsQuery

Ran codegen to update generatedTypes

akikoskinen reviewed (commented):
It confuses me why authentication needs to be mixed with profile fetching. They are unrelated, although dependent concepts. Authentication is a prerequisite for profile loading, but that doesn't  mean that they have to be tightly coupled with each other.

To me it seems that the logic could be something like: 
1. Handle authentication. If not authenticated, ask to authenticate. Handle errors in authentication. Eventually this leads to an authenticated user (or the user goes away, if they can't or don't want to authenticate).
2. After authenticated, start dealing with profile. See if there's a profile present. If not, create one. If yes, show it. Handle errors in fetching.

The execution would go to step 2 only after step 1 has completed successfully. The logic of step 1 would be implemented in it's own module and the logic of 2 would be in another module. ("Module" being a hook or whatever is the correct tool for this).

There shouldn't be any modules (hooks, files, etc.) that has both authentication and profile related stuff, except for one place where there would be something like:

```jsx
if (isAuthenticated()) {
    return <Profile/>;
}
```

NikoHelle commented:
You're right. Would simplify the process if those where separate. I tried to make a small change, but it got bloated and still tried kept the same components.

NikoHelle committed:
(HP-908) Added a new WithAuthChecker wrapper

This will replace authentication checks from Profile page. Profile page should only be used when user is authenticated.

Profile needs the user object, so WithAuthChecker passes it to its child.

If user is not authenticated, user is redirected to login page.

NikoHelle committed:
(HP-908) New useProfileLoadTracker hook

All profile loading scenarios are handled by this hook.

Profile might already be loaded / loading when the hook is initiated which complicates it.

NikoHelle committed:
(HP-908) Use useProfileLoadTracker and WithAuthCheck in App and Profile page

Profile page just uses functions provided by the useProfileLoadTracker hook and does not need to have any logic.

Other than check if loading is was successful and if profile exists.

User object is now passed from parent.

NikoHelle committed:
(HP-908) Added getProfile function to profile context

This way the "profileData && profileData.myProfile" -check is in one place.

NikoHelle committed:
(HP-908) Prevent simultaneous error notification and toast

When error occurs both are shown on Profile page. Removed the toast.

Sending errors to Sentry from components is not the best way.

Better to focus reporting to ProfileContext.

Also shortened "mock.calls.length" checks to "toHaveBeenCalledTimes" in profile context test.

akikoskinen commented on src/profile/hooks/useProfileLoadTracker.ts:
```diff
@@ -0,0 +1,139 @@
+
+export function getLoadFailedStatus(): Status {
+  return STATUS_FAILED;
+}
```
It doesn't seem too useful to have this kind of small helpers that just basically give another name for a simple statement. Even if the new name is useful as an internal detail, I don't see why these should be exported.

NikoHelle replied:
I agree. But they have to be exported for tests. Other option is to export all STATUS_xxx constants, but I think this is cleaner.

akikoskinen replied:
I would prefer exporting the `STATUS_*` constants. They kind of form an enum and are thus grouped to a whole, whereas a bunch of functions are just separate functions without that much connection to each other. But JS/TS is weird to me anyway, so perhaps instead of a neat enum it's customary to create a bunch of oneliner functions :shrug:

NikoHelle replied:
Changed to use STATUS_*

akikoskinen commented on src/profile/hooks/useProfileLoadTracker.ts:
```diff
@@ -0,0 +1,139 @@
+    hasExistingProfile: () => {
+      if (!isLoadComplete(valueRef.current)) {
+        throw new Error(
+          'hasExistingProfile cannot be used before profile load is complete. Check it with isLoadComplete()'
```
```suggestion
          'hasExistingProfile cannot be used before profile load is complete. Check it with isProfileLoadComplete()'
```

NikoHelle replied:
Fixed!

akikoskinen commented on src/profile/hooks/useProfileLoadTracker.ts:
```diff
@@ -0,0 +1,139 @@
+
+export function useProfileLoadTracker(): useProfileLoaderHookReturnType {
+  const profileContext = useContext(ProfileContext);
+  const valueRef = useRef<Status>(STATUS_WAITING);
```
The value's name should be something more explanatory. Perhaps something like `currentStatus`.

NikoHelle replied:
Changed to "statusRef", because there is also "updateStatus", not "updateCurrentStatus" and currentStatusRef.current sounds weird :)

akikoskinen commented on src/profile/components/profile/Profile.tsx:
```diff
@@ -93,60 +39,64 @@ function Profile(): React.ReactElement {
   }
 
   return (
-    <PageLayout title={getPageTitle()} disableFocusing>
```
Should the `disableFocusing` property still be present somewhere in this `Profile` component?

NikoHelle replied:
True. Added to loader in WithAuthCheck too

akikoskinen reviewed (changes requested):
Some smallish notes. Generally looks pretty good.

akikoskinen commented on src/profile/hooks/useProfileLoadTracker.ts:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  return undefined;
+}
```
All these functions are internal implementation details and shouldn't be exported. These are now only used in tests, not in production code. It's better to test the _behaviour_, not _implementation_. So test through the public API.

NikoHelle replied:
Removed implementation tests and unused exports. Added one behavioural test.

In React Apps behavioural tests are quite hard to implement so they are clear and cover all cases. Doable of course, but faking API requests and responses, hooks, dom, user behaviour etc makes tests confusing to read and it takes time to find out if everything is also faked correctly. And all the fake stuff must also be maintained.

Easier to expose and test "safe" parts like resolveChangedStatus() which just returns new status and does not change internal status.

akikoskinen commented on src/profile/hooks/useProfileLoadTracker.ts:
```diff
@@ -0,0 +1,119 @@
+  didProfileLoadFail: () => boolean;
+  reloadProfile: () => Promise<QueryResult>;
+  getStatus: () => Status;
+  getProfile: () => ReturnType<ProfileContextData['getProfile']>;
```
The `getStatus` and `getProfile` functions don't seem to be used in any production code, so can they be removed? If `getStatus` is not needed, perhaps the `STATUS_*` consts and `Status` type don't have to be exported either.

akikoskinen reviewed (changes requested):
`useProfileLoadTracker` still seems to reveal too much of its private parts.

akikoskinen reviewed (approved):
Nice
